test/test_implementation/multi_particle_dist.jl

```diff
@@ -0,0 +1,37 @@
+
+PARTICLE_DIRECTIONS = (Incoming(), Outgoing(), QEDevents.UnknownDirection())
+
+struct TestMultiParticleDist{DT<:Tuple,PT<:Tuple,RT} <: MultiParticleDistribution
+    dirs::DT
+    parts::PT
+    function TestMultiParticleDist(dirs::DT, parts::PT) where {DT,PT}
+        res_type = Tuple{QEDevents._assemble_tuple_types(parts, dirs, SFourMomentum)...}
+        return new{DT,PT,res_type}(dirs, parts)
+    end
+end
+
+#Base.length(d::TestMultiParticleDist) = length(d.parts)
+
+function QEDevents._particles(d::TestMultiParticleDist)
+    return d.parts
+end
+
+QEDevents._particle_directions(d::TestMultiParticleDist) = d.dirs
+
+function Distributions.rand(
+    rng::AbstractRNG, d::TestMultiParticleDist{DT,PT,RT}
+)::RT where {DT,PT,RT}
+    return _groundtruth_multi_rand(rng, d)
+end
```

Comment:
I currently don't like this implementation with the hard return type, but it is necessary for type stability. 

One idea to improve this would involve changing `ParticleSampleable`: 

```julia
abstract type ParticleSampleable{F<:QEDlikeVariate,RET} <: Sampleable{F,Continous}

function rand(rng::AbstractRNG,d::ParticleSampleable{...,RT})::RT where {RT}
    moms = rand_mom(rng,d)
    out = build_out(d,moms)
    return out
end
```
where `rand_mom` is an interface function, and `build_out` can be implemented generically using the type-parameter `RT`. 
The concrete implementation of a distribution would then look like
```julia
struct TestDist{RT} <: ParticleSampleable{...,RT}
    ...
    function TestDist(...)
        return_type = out_type_helper(...)
        new{return_type}(...)
    end
end
```
where `out_type_helper` could be implemented generically as well. It is essentially the current implementation of `Base.eltype`.

@AntonReinhard what do you think? Is it worth it to go in this direction?

Reply:
If I'm not mistaken, `build_out` would (more or less) be taken care of by introducing a `ParticleList` or similar as in https://github.com/QEDjl-project/QEDprocesses.jl/issues/75 .

However, I'm not entirely sure what prevents you from simply doing the `_assemble_tuple_types` in the `rand` call? Since it's only working on the types it should have zero overhead and yield a type-stable function all the same. You would also not need the extra constructor and the `RT` in the Distribution. With your existing wrapper of `Base.eltype` this should be trivial.

Reply:
You are right, of course, if the constructor of a potential `ParticleList` can wrap a tuple or vector of `SFourMomenta` with the respective `AbstractParticleType` and `ParticleDirection` without much overhead. 

The problem with `_assemble_tuple_type` is that it implicitly builds the constructors of `ParticleStateful`, i.e. `ParticleStateful{Direction,Particle,MomType}`. But there is currently no constructor, which allows implicit wraps around a given momentum: something like `ParticleStateful{Dir,Part,MomType}(mom::MomType)`. I played around with it, but such constructors lead to weird implementations, where you need to call type parameters to construct the respective singleton object. 

In e3e02fd117b8d2977b93f5cd07c2636768a9f00b an explicit cast of a vector of four-momenta into a tuple of `ParticleStateful` is added. This makes the usage of return types and implicit casts obsolete and adds only a little overhead if the length of the tuples is not known at compile time. There are even better ways to implement this, but this would involve using generated functions to unroll the tuple construction, which is, IMHO, currently unnecessary.